core/drivers/gic.c

```diff
@@ -200,33 +202,169 @@ static size_t probe_max_it(vaddr_t gicc_base __maybe_unused, vaddr_t gicd_base)
+		/* Ignore matching bits */
+		if (!(BIT32(n) & (grp0 ^ gd->per_cpu_group_status)) &&
+		    !(BIT32(n) & (gmod0 ^ gd->per_cpu_group_modifier)))
+			continue;
+		need_sync = true;
+		/*
+		 * SGI-n differs from primary CPU configuration,
+		 * let's sync up.
+		 */
+		/* Disable interrupt */
+		io_write32(gicr_base + GICR_ICENABLER0, BIT32(n));
+
+		/* Make interrupt non-pending */
+		io_write32(gicr_base + GICR_ICPENDR0, BIT32(n));
+
+		if (BIT32(n) & gd->per_cpu_group_status)
+			grp0 |= BIT32(n);
+		else
+			grp0 &= ~BIT32(n);
+		if (BIT32(n) & gd->per_cpu_group_modifier)
+			gmod0 |= BIT32(n);
+		else
+			gmod0 &= ~BIT32(n);
+	}
+
+	if (need_sync) {
+		io_write32(gicr_base + GICR_IGROUPR0, grp0);
+		io_write32(gicr_base + GICR_IGRPMODR0, gmod0);
+	}
+}
+
+static void gic_legacy_sync_sgi_config(struct gic_data *gd)
+{
+	bool need_sync = false;
+	uint32_t grp0 = 0;
+	size_t n = 0;
+
+	grp0 = io_read32(gd->gicd_base + GICD_IGROUPR(0));
+	for (n = GIC_SGI_SEC_BASE; n <= GIC_SGI_SEC_MAX; n++) {
+		/* Ignore matching bits */
+		if (!(BIT32(n) & (grp0 ^ gd->per_cpu_group_status)))
+			continue;
+		need_sync = true;
+		/*
+		 * SGI-n differs from primary CPU configuration,
+		 * let's sync up.
+		 */
+
+		/* Disable interrupt */
+		io_write32(gd->gicd_base + GICD_ICENABLER(0), BIT(n));
+
+		/* Make interrupt non-pending */
+		io_write32(gd->gicd_base + GICD_ICPENDR(0), BIT(n));
+
+		if (BIT32(n) & gd->per_cpu_group_status)
+			grp0 |= BIT32(n);
+		else
+			grp0 &= ~BIT32(n);
+	}
+
+	if (need_sync)
+		io_write32(gd->gicd_base + GICD_IGROUPR(0), grp0);
+}
+
 void gic_cpu_init(void)
 {
 	struct gic_data *gd = &gic_data;
 
+	if (IS_ENABLED(CFG_WITH_ARM_TRUSTED_FW)) {
+		/*
+		 * GIC is already initialized by TF-A, we only need to
+		 * handle eventual SGI configuration changes.
+		 */
+		if (IS_ENABLED(CFG_ARM_GICV3) &&
+		    io_read32(gd->gicd_base + GICD_CTLR) & GICD_CTLR_ARE_S)
+			gicv3_sync_sgi_config(gd);
+		else
+			gic_legacy_sync_sgi_config(gd);
+	} else {
+		/*
+		 * Non-TF-A case where all CPU specific configuration
+		 * of GIC must be done here.
+		 */
 #if defined(CFG_ARM_GICV3)
-	assert(gd->gicd_base);
+		assert(gd->gicd_base);
 #else
-	assert(gd->gicd_base && gd->gicc_base);
+		assert(gd->gicd_base && gd->gicc_base);
 #endif
 
-	io_write32(gd->gicd_base + GICD_IGROUPR(0), gd->per_cpu_group_status);
+		io_write32(gd->gicd_base + GICD_IGROUPR(0),
+			   gd->per_cpu_group_status);
 
-	/*
-	 * Set the priority mask to permit Non-secure interrupts, and to
-	 * allow the Non-secure world to adjust the priority mask itself
-	 */
+		/*
+		 * Set the priority mask to permit Non-secure interrupts,
+		 * and to allow the Non-secure world to adjust the priority
+		 * mask itself
+		 */
 #if defined(CFG_ARM_GICV3)
-	write_icc_pmr(0x80);
-	write_icc_igrpen1(1);
+		write_icc_pmr(0x80);
+		write_icc_igrpen1(1);
 #else
-	io_write32(gd->gicc_base + GICC_PMR, 0x80);
+		io_write32(gd->gicc_base + GICC_PMR, 0x80);
```

Comment:
Before this change, when `CFG_WITH_ARM_TRUSTED_FW` is enable,
for GICv3 we called `write_icc_pmr(0x80);` + `write_icc_igrpen1(1);`
for GICv2 we called `write_icc_igrpen1(1);` + `io_write32(gd->gicc_base + GICC_CTLR, ...)`
Now its not the case. Is it on purpose?

Reply:
Yes, that's on purpose. Should I do that in a separate commit or perhaps mention it in the commit?

Reply:
It's actually no change in behaviour since `gic_cpu_init()` is normally not called with CFG_WITH_ARM_TRUSTED_FW=y, but with donate SGI it will be needed. I'll make a separate commit and update all the platforms.

Reply:
There are many platforms that enable `CFG_WITH_ARM_TRUSTED_FW` and call `gic_cpu_init()` from `boot_secondary_init_intc()`.

Reply:
Seems tricky to update all platforms in one go. How about deprecating `gic_cpu_init()` and add a new `gic_init_per_cpu()`? `gic_cpu_init()` is kept unchanged with regards to PMR and igrpen1/GICC_CTLR and `gic_init_per_cpu()` is expected by be by all secondary CPUs regardless of `CFG_WITH_ARM_TRUSTED_FW`. I'll submit separate PRs to tidy up the various platforms after this PR.

Reply:
Sounds a good comprise. We should notify platform maintainers to update their implementation so that we can get rid of `gic_cpu_init()` afterward.